[dino.bin] 지하철 step3 리뷰 요청합니다.

kommadot opened the pull request:
## 요약

작업 중간에 피드백 받아보는게 현재 궁금한 부분이나 구현중인 부분에서 많은 도움이 될것 같아서 피드백 요청해봅니다.

Line이 Section을 관리할땐 리스트형식으로 저장하여 리스트의 인덱스 순으로 상행 -> 하행 순으로 저장을 하는 방식으로 step2를 구현하였습니다.
도중에 역 삽입이 발생하면 리스트의 중간 인덱스의 값을 제거하고 새로운 section을 삽입하는 형식으로 구현하였습니다. 현재 step3로 넘어가며 기존 리스트로 저장하던 부분을 db에 저장하도록 변환하는 과도기에 있습니다.
Station은 어느정도 기능적으로 완료가 된 상태입니다.
line과 section은 저장방식이나 구현방식에 대해 고민중에 있습니다

## 질문

1. 컨트롤러가 데이터를 반환할때 Response타입으로 데이터를 반화하고있습니다. 이때 Response타입의 Dto로 변환해주는 역할을 service에서 해야할지 controller에서 해야할지 고민이 됩니다. 어떤 방식이 더 좋은 방식일까요

SimJung committed:
feat : StationController 기능 구현

 - showStations() 메소드를 통해 조회가 가능하도록 구현
 - deleteStation() 메소드를 통해 삭제가 가능하도록 구현

SimJung committed:
feat : 노선 생성 기능 구현

 - Line 클래스 생성
 - LineDao 클래스 생성
 - LineController 클래스 생성
 - 노선 생성 기능 구현 및 테스트 완료

SimJung committed:
feat : 노선 삭제 기능 구현

 - 노선 삭제 기능 구현 및 테스트 완료

SimJung committed:
feat : 노선 조회 기능 구현

 - 노선 전체 목록 조회 기능 구현 및 테스트 완료
 - 특정 노선 조회 기능 구현 및 테스트 완료

SimJung committed:
feat : 노선 수정 기능 구현

 - 노선 수정 기능 구현 및 테스트 완료

kakao-dino-bin committed:
modify: 요구사항 변경에 맞춰 노선 생성 수정

노선 생성시 종점 ID와 거리를 입력받도록 수정

-

kakao-dino-bin committed:
chore: 기타 변경 사항 및 백업 커밋

-

kakao-dino-bin committed:
feat: 구간 추가 기능 구현

노선 추가 후 구간이 추가되며 반환 기능 구현
생성 테스트 수행

-

kakao-dino-bin committed:
feat: 삭제 기능 추가

노선의 역 삭제 기능 추가
테스트 코드의 예외 상황 추가

-

SimJung committed:
feat : DB에 데이터 저장되도록 수정

 - JDBC 의존성 추가
 - Bean 객체 관리 위한 어노테이션 추가
 - Station 정보 DB에서 관리하도록 변경

SimJung committed:
feat : DB 조회하여 역 목록 받아오기

 - 기존의 리스트로 관리하던 역 목록을 DB를 통해 조회할 수 있도록 변경

SimJung committed:
feat : DB 조회하여 역 하나 받아오기

 - DB를 통해 역 하나의 정보를 받아올 수 있도록 변경

SimJung committed:
feat : DB로부터 역 제거하도록 변경

SimJung committed:
feat : StationService 추가

 - interface 및 class 구현하여 컨트롤러에서 서비스 이용하도록 변경

SimJung committed:
feat : 노선 DB 저장 기능 구현

 - Dao에서 관리하던 노선을 DB를 통해 저장할 수 있도록 변경

SimJung committed:
feat : 노선 DB 삭제 기능 구현

SimJung committed:
feat : LineService, SectionController 추가

 - LineController 에서 service를 사용하도록 변경
 - Section 관리를 SectionController에서 하도록 변경

SimJung committed:
refactor : 패키지 분리 및 리팩토링

 - Section 패키지 분리
 - Section 서비스 인터페이스 및 클래스 추가
 - 기타 리팩토링

kakao-dino-bin committed:
modify: line 조회 jdbc연결

이전에 라인 조회는 그저 list에서 가져오는 방식이었지만
db를 연결해서 가져오도록 변경

-

kakao-dino-bin committed:
modify: line 수정 jdbc적용

-

kakao-dino-bin committed:
modify: 구간 추가, 제거, 조회, 라인 일부 수정

구간 추가시 db에 저장되도록 구현
구간 제거, 조회시 db에서 불러와 각 상행 하행이 연결되도록 수정
라인이 시점, 종점을 갖고있도록 수정

-

kakao-dino-bin committed:
fix: 구간 추가, 제거 버그 수정

구간을 추가하거나 제거할때 제대로 안되는 버그를 수정
기능적으로 문제는 발견되지 않음

-

sehan commented:
@kommadot 

안녕하세요. 지하철 미션 리뷰를 하게된 니오입니다. 😊
리뷰를 진행하기 전에 먼저 질문에 대한 답을 먼저 하겠습니다.
 
> ## 질문
> 1. 컨트롤러가 데이터를 반환할때 Response타입으로 데이터를 반화하고있습니다. 이때 Response타입의 Dto로 변환해주는 역할을 service에서 해야할지 controller에서 해야할지 고민이 됩니다. 어떤 방식이 더 좋은 방식일까요

controller 부터 개발을 시작하게 되면 이런 고민을 하게 됩니다. 😊
domain 과 service ( application ) 설계가 controller 를 기점으로 작성되기 때문에 controller 에서 받은 request object 가 service 까지 내려가거나 혹은 domain 까지도 내려갑니다. response object 도 비슷합니다. domain 객체를 우선으로 생각을 해보면 domain 의 method 가 output 으로 data 를 제공하고 그것을 외부에 전달할 Data(DTO) 로 변환하는 작업을 해주는 곳은 1차적으로 현 구조상 service 입니다.  service 를 중간에 두지 않는다면 controller 가 그 역할을 맡게 됩니다. 
 결론적으로 현재 구조상 service 를 중간에 두고 있다면 service 가 우선적으로 그 책임을 가지고 domain 에게 받은 Data 를 변환해서 controller 에게 제공해주는 것이 더 적합해 보입니다. 
하지만, 현재 코드에서 service 는 service 역할을 하지 못하고 있습니다. 단지 dao 를 한번 wrapping 만 하고 있는 상황입니다. 이런 상황에서는 dino 처럼 고민에 빠질 수 밖에 없습니다. service 는 불필요하게 단계만 늘어난 상황이라 service 에서는 dao 의 output 을 bypass 만 하고 있죠. controller 에서 dao 를 직접사용하고 dao 의 output 을 response 객체로 변경해주면 구조가 간결하고 더 직관적이게 됩니다. 그럼 고민 없이 controller 에서 response 로 전달할 DTO 로 변환해주면 되니까요.

sehan commented on src/main/java/subway/station/Station.java:
```diff
@@ -7,6 +7,10 @@
     public Station() {
     }
 
+    public Station(StationRequest stationRequest) {
```
도메인 객체가 Controller 에서 받은 request parameter ( stationRequest ) 객체에 의존하고 있습니다.
도메인의 확장성을 높여주는면에서 의존을 제거 하는것이 좋습니다. 
Service 또는 Controller 에서 Factory 를 사용하여 Station 을 생성하는 방식을 고려해 보세요 ☺️

sehan commented on src/main/java/subway/station/StationController.java:
```diff
@@ -6,25 +6,43 @@
+    @GetMapping(value = "/stations/{stationId}")
+    public ResponseEntity<StationResponse> showStation(@PathVariable Long stationId) {
+
+        return ResponseEntity.ok(new StationResponse(stationService.findOne(stationId)));
```
StationService.findOne() 시 역이 존재하지 않는 경우에 대한 처리가 필요합니다.
역이 없을경우 NPE 가 발생할 위험이 있습니다.

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,87 @@
+    @PostMapping
+    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest lineRequest) {
+        Line newLine = lineService.save(
+                new Line(lineRequest.getName(), lineRequest.getColor(), lineRequest.getUpStationId(), lineRequest.getDownStationId()),
```
Line 생성시 생성인자에 대한 검증이 필요합니다. 도메인 객체에 검증로직을 추가해주세요.
Section 도 동일합니다.

SimJung committed:
modify: section save & delete 로직 수정

 - section 리스트를 관리하는 일급 컬렉션 추가
 - sectinos를 이용하여 관리하도록 하고 기존의 idx를 이용한 삽입과 삭제를 수정
 - 중복 구간 추가 이슈와 구간 추가 시 거리 이슈를 fix함.

SimJung committed:
refactor: Section save, delete 과정 리팩토링

 - sectionServiceImpl의 save, delete 과정을 리팩토링
 - private 메소드를 늘려 구분함
 - Sections unit test 추가

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,62 @@
+            return new Line(id, line.getName(), line.getColor());
+
+        } catch (Exception e) {
+            return null;
```
Dao 에서 exception 발생시 null 을 return 하기 보다는 exception 을 던져 주는 것이 좋습니다. 
하위 layer 에서 exception 을 처리하고 위로 올려주지 않으면 문제 발생시 오류를 찾기가 어려워집니다.

sehan commented on src/main/java/subway/line/LineServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+    public Line save(Line line, Section section) {
+        Line newLine = lineDao.save(line);
+        if (newLine != null) {
+            sectionDao.save(new Section(section.getUpStationId(), section.getDownStationId(), section.getDistance(), newLine.getId()));
```
save() 가 하는 작업은 하나의 트랜젝션으로 묶여야 합니다.
즉, section 저장에 실패하게 되면 앞서 생성한 newLine 도 rollback 이 되어야 합니다.

@Transactional 을 살펴보시고 save() 가 하나의 트랜젝션으로 동작하도록 수정해 주세요.

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,87 @@
+                new Section(lineRequest.getUpStationId(),
+                        lineRequest.getDownStationId(),
+                        lineRequest.getDistance()));
+        if (newLine == null) {
```
null 체크하는 로직이 여러곳에서 발생하고 있습니다. method 에서 null 리턴하는것을 최대한 지양하고 exception 을 활용하거나 null 을 피할수 없다면 Optional 로 처리하는것을 검토해주세요.

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,87 @@
+
+    @GetMapping
+    public ResponseEntity<List<LineResponse>> showLines() {
+        return ResponseEntity.ok(lineService.findAll()
```
domain model 을 DTO 로 변환하는역할을 Converter 로 분리해 내면 코드의 가독성 개선에 도움이 되고 LineResponse 생성자의 signature 가 변경되더라도 코드변경이 Converter 내로 좁혀 지기 때문에 controller 코드가 LineResponse, Line 객체의 변경에 의존하지 않게 됩니다. 해당 부분을 Converter 로 분리해보는 것을 고민해 주세요 😊

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,87 @@
+        for (Section section : sections) {
+            stationIds.add(section.getUpStationId());
+            stationIds.add(section.getDownStationId());
+        }
```
도메인 로직으로 볼 수 있는 코드입니다. controller 에서 도메인 로직을 제거하고 적합한 곳으로 이동시켜 주세요

kommadot replied:
이부분을 도메인이나 컨버터를 이용해 변환하려고 해봤는데, Service 빈을 참조하기 때문에 도메인이나 컨버터에선 변환이 불가능할거같아 우선 Service로 책임을 옮겼습니다. 다른 방법이 있을까요

sehan commented on src/main/java/subway/line/LineServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public boolean deleteById(Long lineId) {
+        return lineDao.deleteById(lineId) != 0;
```
line 삭제시 라인에 포함된 section 들도 삭제되어야 합니다.  그리고 하나의 트랜젝션으로 동작되어야 합니다.

sehan commented on src/main/java/subway/section/SectionServiceImpl.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    public boolean saveSection(Long lineId, Section section) {
+        List<Section> sections = getSectionsByLineId(lineId);
```
도메인 객체가 협력을 할 수 있도록 해주세요. 지금은 Data 를 꺼내어 로직을 구현하고 있습니다. 해당 Data 를 상태로 관리하고 있는 도메인 객체에게 로직을 넘겨주고 객체가 협력하도록 작성해주세요. 로직을 최대한 도메인 객체로 옮겨 보세요 😊

지금 도메인 layer 가 얇고 서비스 layer(application layer) 가 두꺼운 상태입니다.  
도메인 layer 로 로직을 최대한 옮겨서 서비스 layer 의 두께를 줄여주세요

kommadot replied:
현재 서비스 레이어의 대부분이 dao 혹은 다른 service접근인데 이를 어떻게 더 줄일 수 있을까요?

sehan replied:
비지니스 로직을 Service layer 에서 구현하고 있기 때문에 지금 방식으로는 줄이기가 어렵습니다. 

현재코드는 Service layer 에서 섹션을 추가하기 위해서 dao 로 필요한 데이터를 가져와서 추가에 필요한 로직들을 수행하고 있습니다. 이렇게 Service layer 에 풀어쓰는 방식을 transactin script 라고 합니다. 

객체간 협업을 으로 구성되었다면 라인에 섹션을 추가하기 위해 dao 로 Line 객체를 가져옵니다. Line 객체에 섹션을 추가해달라고 요청합니다. Line 객체는 내부로직에 따라 섹션을 추가합니다.  이와 같은 흐름이라면 아래와 같은 형태로 Service 코드가 작성되어야 합니다.
```
LineService {

void addSection( long lineId, Station up, Station down, long distance ){
     Line line = dao.findOne(lineId);
     line.addSection(up, down, distance);
     sectionDao.save(line.getSections());   또는 dao.save(line);
}
}
```
비지니스 로직이 제거되고 객체간 협력(오케스트레이션) 을 통해서 구현된 코드는 위와 같은 방식으로 나와야 합니다.

sehan reviewed (changes requested):
안녕하세요. 지하철 미션을 함께 하게된 니오입니다. 😊
리뷰 많이 기다렸을 텐데 너무 늦어져서 죄송합니다.  🙏🙏🙏

전체적인 기능구현과 dao / service 구성을 잘 해주셨습니다. 👍 
다만, 몇가지 개선할 부분이 있어서 수정요청드립니다.
현재 비지니스 로직이 대부분 Service 객체에 구현되어 있습니다. 도메인은 얇고 서비스가 두꺼운 상황입니다. 이 부분에 대한 집중적인 개선이 필요합니다. 
 그 외로는 몇가지 코멘트를 남겨 놨습니다.  코멘트 확인 부탁드립니다.

kakao-dino-bin committed:
refactor: 1차 리팩토링, Controller의 역할 최소화

Controller가 데이터 변환등 하는 일을 최소화

-

kakao-dino-bin committed:
refactor: 패키지 변경

패키지를 도메인에 따른 패키지가 아닌 Controller, Service, Dao, Domain, Dto 별로
변경

-

kakao-dino-bin committed:
refactor: 예외, 검증 추가 및 팩토리 컨버트 추가

예외 클래스와 검증 구문을 추가해서 검증 후 예외가 호출되도록 수정
팩토리와 컨버트 클래스를 추가

-

kakao-dino-bin committed:
refactor: Tranactional추가 및 리팩토링

-

sehan commented on src/main/java/subway/service/LineServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+        return lineDao.updateAll(line) != 0;
+    }
+
+    public List<LineResponse> findAllResponse() {
```
이 메소드는 LineService.findAll() 과 LineConvertor.convertLines() 조합으로 controller 에서 풀어낼 수 있을 것으로 보이네요.
꼭 필요한 메소드가 아니라면 제거해주는게 좋겠습니다. 
더욱이 LineService 인터페이스로 선언되는 메소드라면 상세구현체들의 모두 해당 메소드를 구현해야 해서 메소드로 추가할 필요가 있는지 더 고민할 필요가 있습니다.

sehan commented on src/main/java/subway/service/LineServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+                .collect(Collectors.toList());
+    }
+
+    public LineResponse findOneResponse(Long lineId) {
```
findAllResponse 와 같이 이 메소드를 왜 추가했을까? 하고 보니 LineResponse 에 station 목록을 추가하기 위함이군요.
Response 의 요구사항이 Service 에 직접적으로 추가된 것으로 보입니다. 요구사항을  Service 에 직접구현 할 수도 있지만 그 에 앞서 도메인에서 요구사항을 충족시킬 수는 없는지 고민해볼 필요가 있겠습니다.

그런데 Line 객체를 보니 도메인 이라기 보다는 DB table 과 1:1 로 매핑되는 entity 역할 을 하고 있습니다. 
Line 을 도메인 객체로 바라봤다면 아래와 같은 멤버와 메소드를 가지고 제공 할 수 있을거 같습니다.
```
class Line {
 private Long id;
 private String name;
 private String color;
 private List<Station> stations;

 public Station getUpStation();
 public Station getDownStation();
}
```

도메인 객체가 위와 같다면 Line 으로 LineResponse 를 만들 수 있을겁니다. 

도메인객체와 엔티티객체를 분리하는것이 쉬운일은 아니지만 앞서 수행했던 자동차경주/로또 미션을 생각해본다면 도메인 설계를 왜 진행했는지 한번 되새겨 볼 필요가 있습니다. 이 부분은 어려운 주제 이니 가능하다면 시도 해보시기 바랍니다. 시간적 제약으로 어려울 수도 있을거 같아요 ☺️

sehan commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -0,0 +1,61 @@
+
+    private final LineService lineService;
+
+    public LineController(LineService lineService) {
```
controller 가 아주 멋지게 다이어트 성공했네요 👍 👍 👍

sehan commented on src/main/java/subway/dto/LineResponse.java:
```diff
@@ -1,23 +1,38 @@
     private String name;
     private String color;
-    private int extraFare;
     private List<StationResponse> stations;
```
LineResponse 가 상황에 따라 stations 정보가 있을수도 있고 없을수도 있다면 용도에 맞게 Response Data 객체를 분리하는 것이 좋습니다. LineResponse 를 사용하는 측에서는 어느 상황에 stations 가 포함되는지 알 수 없어서 혼돈을 주기 때문입니다.
그리고 client 코드를 작성시 상황에 따른 방어코드를 삽입하게 만들기 때문입니다.

sehan commented on src/main/java/subway/service/LineServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.stream.Collectors;
+
+@Service
+public class LineServiceImpl implements LineService {
```
implement 한 상세구현체에 @Override 어노테이션이 모두 제거가 되어 있네요?
상세구현체에서 추가된 메소드인지 override 된 메소드인지 구분을 위해 붙여 주는게 좋습니다.
인텔리제이에서 자동으로 붙여주는데 의도적으로 삭제한거 같아서 코멘트 남깁니다.

sehan commented on src/main/java/subway/controller/SectionController.java:
```diff
@@ -0,0 +1,34 @@
+
+    @PostMapping("/{lineId}/sections")
+    public ResponseEntity createSection(@PathVariable Long lineId, @RequestBody SectionRequest sectionRequest) {
+        if (!sectionService.saveSection(SectionFactory.getSection(sectionRequest, lineId))) {
```
saveSection 에서 이미 Section 이 존재하면 false 를 return 해주고 있습니다. 

Response Error 처리를 위해 spring 에서는 application 에서 발생하는 exception 들을 한곳에 처리할 수 있도록 해주는 기법들을 제공합니다.  현재 미션에서는 사용하고 있지 않지만 위와 같은 상황에서도 custom exception 을 던지고 처리하는 것이 spring 의 error handling 방식을 적용하기에 더 좋습니다. 

그리고 위 에러상황은 검증로직에 이미 section 이 존재해서 생성할 수 없다는 것이기 때문에 500 보다는 bad request 가 더 접합해 보입니다.  500 은 서버에서 처리하지 못하는 오류가 발생했을 경우에만 제공하는 것이 좋습니다.

sehan commented on src/main/java/subway/service/SectionServiceImpl.java:
```diff
@@ -0,0 +1,126 @@
+        return checkAndAddSection(section);
+    }
+
+    private boolean checkAndAddSection(Section section) {
```
이 부분에 많은 로직이 포함되어 있습니다. 아마도 잘 인지하고 있을거 같네요 ☺️
Section 의 데이터 업데이트 뿐 아니라 Line 의 upStationId, downStationId 까지 업데이트 하는 로직도 함께 있어서 더 커진 면도 있어보입니다.

upStationId, downStationId 를 DB LINE 테이블의 필드로 저장할 수도 있지만 저장하지 않는 방법도 있습니다. 즉, 현재는 upStationId 와 downStationId 를 LINE 테이블에 추가하는 역정규화를 사용했는데 미션의 의도는 그 부분을 도메인로직에서 풀어내 줄것을 요구하고 있지 않나 하는생각이 듭니다.  효율이나 성능면에서 역정규화를 사용하는 장점도 분명히 있습니다. 그렇게 하는것은 Entity 중심의 설계 방식입니다.  나쁘다는건 아니고 현재 미션에서 요구하는건 객체지향설계의 방향성을 지향하고 있어서 입니다.

앞서 이야기 했던 도메인설계 측면에서 이부분도 고민을 해주시면 좋겠습니다.

sehan commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,61 @@
+            Long id = simpleJdbcInsert.executeAndReturnKey(parameters).longValue();
+            return new Line(id, line.getName(), line.getColor(), line.getUpStationId(), line.getDownStationId());
+        } catch (Exception e) {
+            throw new AlreadyExistData();
```
custom exception 을 사용하는 것은 좋습니다. 👍 
하지만, try / catch 에서 발생하는 exception 을 모드 AlreadyExistData() 로 덮어 버려서 진짜 오류내용은 사라지고 AlreadyExistData() 만 남게 되었네요.

어떤 Error 인지 판단이 어렵고 caller 에서 throws 를 하고 싶다면 RuntimeException 을 사용하고 cause 를 포함해서 위로 던져 주는것이 좋습니다. 추후에 exception 별 처리가 명확해 지면 그때는 명확하게 exception 던지도록  처리해줘야 합니다.

마지막으로 AlreadyExistData 객체에는 @ResponseStatus(HttpStatus.BAD_REQUEST) 가 붙어 있습니다. 이건 controller 즉, UI 에 속하는 영역입니다. Dao 에 UI 의존이 발생하기 때문에 사용하면 안됩니다. 
layered architecture 상에서는 밑에 있는 layer 에서 상위 layer 로 올려준 exception 들을 최상위 layer 에서 exception 을 아 변환하여 UI 로직에 맞게 처리하는 것이 좋습니다. 

spring 에서의 exception-handling 은 아래 문서를 한번 참고해보세요
https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc

kommadot replied:
하위 레벨에서 throw를 해도 controller advice가 잘 캐치를 하긴 하는데, 이를 상단에서 try catch문으로 다시 핸들링 해줘야하나요?

sehan commented on src/main/java/subway/service/LineServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Transactional
+    public boolean deleteById(Long lineId) {
+        if(lineDao.deleteById(lineId) == 0){
+            return false;
+        }
+        sectionService.deleteSectionByLineId(lineId);
```
line 을 삭제할때 연관된 entity 들이 존재하면 하위(자식) entity 들을 먼저 삭제해야 합니다.
지금 스키마에는 FK 나 식별을 위한 제약사항이 걸려 있지 않지만 이것들이 존재하면 상위(부모) entity 삭제시 오류가 발생합니다.
line 단건 삭제시 sql update 가 반영된 rowboat 를 체크 하기 보다는 exception 이 없다면 삭제된것으로 신뢰할 수 있도록 코드를 만들면 분기문을 제거 할 수 있습니다.

sehan reviewed (changes requested):
1차 피드백 내용을 아주 잘 반영해 주셨습니다 👍 👍 👍 

몇가지 보완이 필요한 내용들을 코멘트에 자세히 남겨두었습니다.
간단히 요약하면
* exception 처리방법에 개선이 필요합니다.
* domain 이라고 만들어 놓은 부분들은 실제로 entity 에 가깝다고 할 수 있습니다.  도메인 설계 관점에서 한번 더 고민해주세요 

한번 더 수정요청 드립니다. 화이팅~

kakao-dino-bin committed:
refactor: 사용안하는 메소드 제거, 도메인 역할 강화

LineResponse를 상세 조회일때, 목록조회일때의 반환값을 분리
Override어노테이션 추가

-

kakao-dino-bin committed:
refactor: 예외처리를 Controller Advice에서 처리하도록 수정

예외처리를 개별로하고 return false등으로 처리했는데 Controller advice에서 처리하도록 변경

-

kakao-dino-bin committed:
refactor: Line, Section을 더 도메인 답게 수정

Line이 Sections를 가지고있으며, Sections가 Stations를 갖고있어 서로 더욱 협력하도록
협력으로 인해 saveSection의 두께를 다이어트

-

kakao-dino-bin committed:
refactor: line service의 delete분리

delete가 한메소드에 너무 많은 라인을 갖고있어서 분리

-

kakao-dino-bin committed:
refactor: findOneResponse 제거

Line이 이미 역 정보를 다 갖고 있으므로 Convertor가 데이터 수집 가능

-

kakao-dino-bin committed:
refactor: 코드 컨벤션 적용 및 미사용 생성자 제거

kakao-dino-bin committed:
refactor: 사용 안하는 라이브러리, 파라미터 제거

-

kommadot commented:
## 요약
말씀하신대로 Service를 최대한 다이어트 시켰습니다.
도메인에게 로직들에 대한 책임을 부여하는 식으로 구현하고
service는 dao를 호출하는 구조로 최대한 사용하려 하였습니다.

## 질문
1. service를 최대한 가볍게 하다보니 domain이 뚱뚱해졌는데 어디까지가 상정내 인지 감이 잘 안잡힙니다.

sehan commented:
> ## 질문
> 1. service를 최대한 가볍게 하다보니 domain이 뚱뚱해졌는데 어디까지가 상정내 인지 감이 잘 안잡힙니다.

최소한 미션을 수행하는 중에는  domain 은 뚱뚱한게 미덕입니다. ☺️
web , service , repository layer 없이 domain 만 가지고 console 용 지하철 노선도 어플리케이션을 만든다고 생각해 보면 좋겠습니다.
domain 이 뚱뚱할 수록 중복으로 구현해야 하는 비지니스 로직들이 줄어듭니다. 
domain 이 허술할 수록 어플리케이션을 구현하면서 버그발생이 늘어나고 비지니스 로직이 달라질 수도 있습니다.

물론 이건 domain 을 가지고 다양하게 활용하는 이상적인 환경에서 잘 만들어 놓은 domain 이 그 힘을 발휘합니다.
현업에서 마주하는 많은 domain 들은 하나의 어플이케이션에서만 활용되는것이 거의 99%에 달하고 개편시에 이전코드를 전면 재개발하게 됩니다. 재개발 비용이 아주 크죠.  
 domain 이 각 layer 에 독립적으로 잘 구현되어 있다면 전면개편을 들어가더라도 기존 domain 을 개선해서 재사용해보려는 시도를 해볼 수 있을 거 같습니다. 이것 또한 참 이상적인 상황이긴 합니다.  그래서 질문하신 것처럼 어디까지 해야 하나? 고민이 들게 됩니다.

장인정신으로 코드를 빚어낸다는 마음가짐으로 domain 을 빚어주시기를 바랄 뿐입니다.

sehan commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -39,9 +37,9 @@ public LineController(LineService lineService) {
-        return ResponseEntity.ok(lineResponse);
+    public ResponseEntity<LineResponseWithStation> showLine(@PathVariable Long lineId) {
+        LineResponseWithStation lineResponseWithStation = lineService.findOneResponse(lineId);
+        return ResponseEntity.ok(lineResponseWithStation);
```
👍

sehan commented on src/main/java/subway/service/LineServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+    public void deleteById(Long lineId) {
+        sectionService.deleteSectionByLineId(lineId);
+        if (lineDao.deleteById(lineId) == 0) {
+            throw new DeleteImpossibleException();
```
삭제가 되지 않을 경우를 rowcount 로 확인하고 있습니다. sql 을 직접 사용하다보니 방어코드 성격이 강해보입니다.
lineDao.deleteById() 호출시 exception 발생하지 않으면 정상적으로 삭제가 되었다고 신뢰할 수 있는 코드를 만드는 것이 좋습니다.
그렇지 않으면 삭제와 유사한 동작을 하는 insert / update 시 에도 정상적으로 sql 이 동작했는지 rowcount 를 확인해야 합니다.

batchInsert 와 같이 다량의 row 에 대해 insert / update / delete 할 경우 rowcount 확인이 유효할 수는 있으나 단건의 경우 row count 체크 보다는 delete 매소드는 무조건 성공하도록 하고 문제가 발생시 exception 으로 처리하는 것을 추천합니다.

sehan commented on src/main/java/subway/controller/AppGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,47 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import subway.exception.*;
+
+@RestControllerAdvice
```
Global exception handler 로 error 응답처리를 잘 정리하셨네요 👍 👍 👍 

여기서 error response message 포멧을 잘 설계해 준다면 더 구조적인 error response message 가 될거 같습니다. 
추후에 ErrorAttributes 에 대해서 한번 살펴보세요 😊

sehan commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,62 @@
+        this.sectionDao = sectionDao;
+    }
+
+    public Line save(Line line, Section section) {
```
Line 과 기본 Section 을 함께 저장하는 시도 좋네요 😊
코드는 간결해지지만 단점은 Line 과 Section 을 함께 저장해야 한다는 요구는 LineController 에서 데이터를 한번에 받고 있어서 그 요구사항이 Dao 까지 그대로 전달되어 만들어진 것으로 보입니다. Line 저장시 Section 을 저장하지 않아도 된다고 요구사항이 변경된다면 Dao 까지 수정해야하는 일이 발생합니다. 
 다시말하면 비지니스 로직이 Dao 에 영향을 주어 save() 가 비지니스 로직에 의존적이 된것으로 보입니다. LineDao.save(line) SectionDao.save(section) 두개의 operation 을 Service 에서 트랜젝션으로 묶어주면 즉, Dao 를 협력시키면 문제를 해결할 수 있기에 save(Line, Section) 하는 시도는 좋지만 비지니스 로직에 의존하는 단점이 있습니다.

sehan commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,62 @@
+
+    public List<Line> findAll() {
+        String sql = "select id from LINE";
+        return jdbcTemplate.query(sql, (rs, rowNum) -> findOne(rs.getLong("id")));
```
Line 목록 조회시 Line 별로 Station 목록을 추가해 줄 것인가? 그 고민에 RowMapper 에서 findOne 을 사용한 것으로 생각되네요.
slack DM 으로 일단 성능은 뒤에 생각해보자고 한말 때문에 그런것인가요? 😅
production 에서는 절대 사용해서는 안되는 방법입니다. 혹시나 해서 다시 이야기 합니다. 

요구사항을 다시 생각해보면 개별 lines/1 조회 시에만 Station 목록을 제공하고 있습니다.  실제로 LineController 에서 Response DTO 로 변경할때도 사용하지 않고 있습니다.  잘 동작하는지 확인하려고 작성된 코드라면 👌

sehan commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,62 @@
+    public Line findOne(Long lineId) {
+        String getLineSql = "select * from LINE where id = ?";
+        Line line = jdbcTemplate.queryForObject(getLineSql, (rs, rowNum) -> new Line(rs.getLong("id"), rs.getString("name"), rs.getString("color")), lineId);
+        Sections sections = sectionDao.getSectionsByLineId(lineId);
```
Dao 내에서 한번에 Line 객체의 Section 까지 로딩한 시도도 아주 좋습니다. 👍 👍 👍 
아쉬운점은 SectionDao 의존을 가지고 있다는 것입니다.
SectionDao 의존을 제거하면 더 좋겠지만 그건 step4 에서 한번 고민해 보면 좋겠습니다

sehan commented on src/main/java/subway/dao/SectionDao.java:
```diff
@@ -0,0 +1,62 @@
+                rs.getLong("down_station_id"),
+                rs.getInt("distance"),
+                rs.getLong("line_id")), lineId));
+        List<Long> ids = sections.getStationIds();
```
Sections 를 Dao 에서 적극적을 활용하고 있습니다. 그러다 보니 Dao 에서 필요한 매소드 ( getStationIds() ) 가 Sections 에 추가된 걸로 보입니다. Sections 는 도메인 객체입니다.  도메인 내에 있는 객체들간의 협업을 위해 설계가 되어야 합니다. Dao 내 로직 구현을 위한 메소드 추가로 보여서 의견드립니다.

sehan commented on src/main/java/subway/domain/Line.java:
```diff
@@ -0,0 +1,53 @@
+    private Long id;
+    private final String name;
+    private final String color;
+    private Sections sections;
```
private Long upStationId;
    private Long downStationId;

를 제거하고 Sections 가 멤버로 들어오니 좀 더 객체 스러워 졌습니다. 👍 
이제 Line 객체가 좀 더 그에 맞는 책임을 가지게 되면 더 좋아 지겠네요

sehan commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,131 @@
+
+import java.util.*;
+
+public class Sections {
```
Sections 에 station 목록 상태를 관리하니 점점 더 도메인 스러워 지고 있습니다. 👍

sehan commented on src/main/java/subway/service/SectionServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    @Override
+    @Transactional
+    public void saveSection(Section section) {
+        if (sectionDao.existSection(section)) {
```
이제 Line 이 좀 도메인객체 스러워 졌으니 Line 에게 물어봐도 되지 않을까요?

sehan commented on src/main/java/subway/service/SectionServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private void checkAndAddSection(Section section) {
+        Sections sections = sectionDao.getSectionsByLineId(section.getLineId());
```
Line 이 sections 를 관리하고 있으니 line 에게 물어봐도 되지 않을까요?

sehan commented on src/main/java/subway/service/SectionServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+        if (!sections.isCanSaveSection(section)) {
+            throw new IllegalStationException();
+        }
+        save(section);
```
저장부분은 section 이외에 영향받는 section 까지 저장해야 하니 이걸 어찌 해야 할까요?

한가지 팁은 line 에 모든 section 목록이 있기에 line 이 관리하는 sections 의 변경사항을 db 에 반영하면 되지 않을까요?
말은 쉽지만 변경사항 관리가 쉽지는 않습니다 ☹️  도전과제로 드립니다.

sehan reviewed (approved):
피드백을 아주 잘 반영해 주셨습니다. 💯 💯 💯 

exception 을  global exception handler 로 아 처리하니 controller 가 아주 깔끔해 졌습니다. 👍 
dao 개선 부분도 아주 인상적입니다.

이번 step 은 merge 하겠습니다. 아주 잘 해주셨습니다.
다음 step 에서 고민해 볼 내용들을 코멘트에 남겨 두었습니다. 코멘트 확인 부탁드려요 ☺️
